src/support_code_library_builder/types.ts

```diff
@@ -35,7 +35,7 @@ export type TestCaseHookFunction<WorldType> = (
 export type TestStepHookFunction<WorldType> = (
   this: WorldType,
   arg: ITestStepHookParameter
-) => void
+) => any | Promise<any>
```

Comment:
`any` implies we expect an interesting return value. Do we look at the return value of a hook? If not, I suggest `void | Promise<void>` is better.

Reply:
Good question.

It's not explicitly documented (per https://github.com/cucumber/cucumber-js/issues/1312#issuecomment-1016807156 which we need to address) but the return value can be "skipped" or "pending" as with regular steps.

It would be good to add this explicitly to the typings but I'm wary of suddenly causing compilation issues for people that e.g. are using chain promises and implicitly returning the result.

I'll do a little experimenting.

Reply:
`string | void | Promise<string | void>` won't work?

Reply:
(It doesn't really make sense to `return 'pending'` from a `BeforeStep`, so we should reconsider that, which I think was just an accident of the implementation.)

Reply:
> `string | void | Promise<string | void>` won't work?

My thought was around user code like:

```
BeforeStep(() => {
  return somethingAsyncThatReturnsPromise()
})
```

Where they are returning the promise so cucumber awaits it, but it resolves to some value that isn't a string so it fails the compilation, and now they have to refactor. Maybe less relevant now with `async/await` being so prevalent though.

Reply:
> We may use `any` in a first time, implementing a deprecation notice when not being an expected return type, then fix it in a version after that?

Yeah that sounds like a good plan. We need to rationalise this, plus the return types for the other hooks and steps, which I think should happen in a major release.

Reply:
Just FYI, you can use a concrete string value as a type in Typescript, so it could be `"skipped" | "pending" | void` which is narrower than `string | void`.

I see your point about keeping it loose though David, consider also people this:

`BeforeStep(somethingAsyncThatReturnsPromise)`

Reply:
Yep another good example.

What do we think then to merging this and raising something new to fix all these types properly for 9.0.0?

Reply:
Good plan. 👍

Reply:
New issue here https://github.com/cucumber/cucumber-js/issues/2040